File: src-ruby/lib/tlg0_reader.rb
require_relative 'tlg_reader'
require_relative 'tlg5_reader'
require_relative 'tlg6_reader'
require 'rubygems'
require 'RMagick'

class Tlg0Reader < TlgReader
  MAGIC = "\x54\x4c\x47\x30\x2e\x30\x00\x73\x64\x73\x1a"

  def read_stream(file)
    assert_magic(file)
    _raw_data_size = file.read(4).unpack('<L')[0]
    magic2 = file.read(11)
    file.seek(-11, 1)
    image = reader_from_magic(magic2).new.read_stream(file)
    process_chunks(file)
    image
  end

  private

  def process_chunks(file)
    until file.eof?
      chunk_name = file.read(4)
      chunk_size = file.read(4).unpack('<L')[0]
      chunk_data = file.read(chunk_size)
      process_tag_chunk(chunk_data) if chunk_name == 'tags'
    end
  end

  def process_tag_chunk(chunk_data)
    until chunk_data.empty?
      key, chunk_data = extract_string(chunk_data)
      _, _, chunk_data = chunk_data.partition('=')
      value, chunk_data = extract_string(chunk_data)
      _, _, chunk_data = chunk_data.partition(',')
      puts 'Tag found: ' + key + ' = ' + value
    end
  end

  def extract_string(raw)
    size, _colon, raw = raw.partition(':')
    size = size.to_i
    value = raw[0..(size - 1)]
    raw = raw[size..-1]
    [value, raw]
  end

  def assert_magic(file)
    fail 'Not a TLG0 file ' if file.read(11) != MAGIC
  end

  def reader_from_magic(magic)
    return Tlg5Reader if magic == Tlg5Reader::MAGIC
    return Tlg6Reader if magic == Tlg6Reader::MAGIC
  end
end

File: src/Tlg0Reader.cc
#include <cstring>
#include <string>
#include <iostream>
#include <memory>
#include <stdexcept>
#include "Tlg0Reader.h"
#include "Tlg5Reader.h"
#include "Tlg6Reader.h"

namespace
{
	std::string extract_string(char *&ptr)
	{
		int length = 0;
		while (*ptr >= '0' && *ptr <= '9')
		{
			length *= 10;
			length += *ptr++ - '0';
		}

		if (*ptr != ':')
			return "";
		ptr ++;

		std::string value;
		for (int i = 0; i < length; i ++)
			value += *ptr ++;

		return value;
	}

	void process_tag_chunk(std::unique_ptr<char> chunk_data, size_t chunk_size)
	{
		char *start = chunk_data.get();
		char *ptr = start;
		while (ptr < start + chunk_size)
		{
			auto key = extract_string(ptr);
			ptr ++;
			auto value = extract_string(ptr);
			ptr ++;
			std::cout << "Tag found: " << key << " = " << value << std::endl;
		}
	}

	void process_chunks(std::ifstream &ifs)
	{
		char chunk_name[4];
		while (ifs.read(chunk_name, 4))
		{
			uint32_t chunk_size;
			ifs.read((char*) &chunk_size, 4);

			auto chunk_data = std::unique_ptr<char>(new char[chunk_size]);
			ifs.read(chunk_data.get(), chunk_size);
			if (std::strncmp(chunk_name, "tags", 4) == 0)
			{
				process_tag_chunk(std::move(chunk_data), chunk_size);
			}
		}
	}
}

const std::string Tlg0Reader::get_magic() const
{
	return std::string("\x54\x4c\x47\x30\x2e\x30\x00\x73\x64\x73\x1a", 11);
}

const Image Tlg0Reader::read_raw_data(std::ifstream &ifs) const
{
	uint32_t raw_data_size;
	ifs.read((char*) &raw_data_size, 4);

	auto reader = AbstractTlgReader::choose_reader(ifs);
	auto ret = reader->read_raw_data(ifs);

	process_chunks(ifs);

	return ret;
}
